### PR TITLE
refactor(ui): readme fetch and rendering

### PR DESCRIPTION
**Pull request by xla**
This improves the way the Readme is fetched and rendered and gets rid of a lot of nested conditionals. Achieved through the extension of Remote with an optional context it can create, which is sub-optimal but the best approach at the moment. We also introduce store chaining, where one store is dependent on the successful remote fetch of another.

Follow-up to #311

### CODE REVIEW

File: proxy/src/coco.rs [hunk 1]
```diff
@@ -137,6 +137,8 @@ pub struct Blob {
     pub content: BlobContent,
     /// Extra info for the file.
     pub info: Info,
+    /// Absolute path to the object from the root of the repo.
+    pub path: String,
```
Comment (cloudhead) [inline proxy/src/coco.rs:141]: Lol I wanted to add this, but stopped myself.

File: ui/DesignSystem/Component/Remote.svelte [hunk 2]
```diff
@@ -1,15 +1,24 @@
 <script>
   import * as remote from "../../src/remote.ts";
 
+  import WithContext from "./WithContext.svelte";
```
Comment (rudolfs) [inline ui/DesignSystem/Component/Remote.svelte:4]: Wrapper inception :)

**Comment by cloudhead** [comment]
Nice. I can't see where binary readme content is handled though?

**Comment by xla** [comment]
> Nice. I can't see where binary readme content is handled though?

@cloudhead What could we sensibly do with a binary readme? Treated same as a non-existing one, is there a plan for that case?

**Comment by cloudhead** [comment]
> > Nice. I can't see where binary readme content is handled though?
> 
> @cloudhead What could we sensibly do with a binary readme? Treated same as a non-existing one, is there a plan for that case?

Where is that code though? I can't find it.

File: ui/DesignSystem/Component/Remote.svelte [hunk 3]
```diff
@@ -1,15 +1,24 @@
...
+  export let context = null;
 </script>
 
 {#if $store.status === remote.Status.NotAsked}
   <slot name="not-asked" />
 {:else if $store.status === remote.Status.Loading}
   <slot name="loading" />
 {:else if $store.status === remote.Status.Success}
-  <slot data={$store.data} />
+  {#if context && typeof context === 'string'}
```
Comment (sarahscott) [inline ui/DesignSystem/Component/Remote.svelte:15]: Curious about this `typeof context === 'string'` guard - what are the other possibilities for the `context`, and why would we not use a `WithContext` wrapper for them?

File: ui/Screen/Project/Source.svelte [hunk 4]
```diff
@@ -48,33 +47,21 @@
...
-  let readmePath = null;
+  const getRevision = current => {
+    return current !== "" ? current : metadata.defaultBranch;
+  };
 
-  $: readme = null;
-  $: if ($objectStore.status === remote.Status.Success) {
-    if ($objectStore.data.info.objectType === TREE) {
-      readmePath = findReadme($objectStore.data);
+  $: readmeStore = readme(id, getRevision($currentRevision));
```
Comment (sarahscott) [inline ui/Screen/Project/Source.svelte:56]: Nice, this all looks really slick!

File: ui/src/source.ts [hunk 5]
```diff
@@ -224,3 +221,28 @@ export const findReadme = (tree: Tree): string | null => {
...
   return null;
 }
+
+export const readme = (
+  projectId: string,
+  revision: string,
+): Readable<remote.Data<Readme | null>> => {
+  const readme = remote.createStore<Readme | null>();
+
+  remote
```
Comment (sarahscott) [inline ui/src/source.ts:231]: This bit is a little 🤯 for me; can you provide a quick tldr of what the chain is doing here?

File: ui/DesignSystem/Component/Remote.svelte (hunk 3 above)
Comment (xla) [reply ui/DesignSystem/Component/Remote.svelte:15]: We use that prop to name the context variable provided, so we want to make sure it's only used as such, happy to see alternatives.

**Comment by xla** [comment]
> Where is that code though? I can't find it.

Good catch, thought I accounted for it. Addressed in 5d03559.

File: ui/src/source.ts (hunk 5 above)
Comment (xla) [reply ui/src/source.ts:230]: `chain` ties one store to another, advancing their loading state in lockstep, and setting the error status if there was one in the dependent store. Once the dependent store has transitioned to `Success` the promise is resolved, that we use here to see  if the loaded object is a tree, search for a readme and load that readme.

**Pull request merged**